Migration Tutorial: From Public to Cloud (#44)
appuio/appuio-cloud-docs: docs/modules/ROOT/pages/tutorials/migration.adoc

= Migration Tutorial: From APPUiO Public to APPUiO Cloud

> This document provides an overview of all the tasks required to move your applications from APPUiO Cloud to APPUiO Public.

Reviewer: :D
```suggestion
This document provides an overview of all the tasks required to move your applications from APPUiO Public to APPUiO Cloud.
```

Author: My eyes! My eyes!!

== Main Differences

Before starting the migration of your applications from APPUiO Public to APPUiO Cloud, please check the xref:explanation/differences-to-public.adoc[document explaining all the major differences] between both platforms.

== DNS

Customers transferring their domains from APPUiO Public to APPUiO Cloud should be aware of the following:

> . Lowering your TTLs (Time To Live) 24 hours ahead of time will reduce the amount of time that your change takes to propagate. You can limit downtime by lowering that value, so the caching servers will check back more frequently. Typically, 60 seconds (1 minute) is a good idea.

Reviewer: ```suggestion
. Lowering your TTLs (Time To Live) 24 hours ahead of time will reduce the amount of time that your change takes to propagate. 
  You can limit downtime by lowering that value, so the caching servers will check back more frequently. 
  Typically, 60 seconds (1 minute) is a good idea.
```

Author: My eyes! My eyes!!

+
IMPORTANT: Remember to set your TTL back to the previous value after the migration.

> . There are different `CNAME` resource records to point to depending on which {zone} you migrated the application to; please refer to xref:zones.adoc[the {zone}s] reference document, where the corresponding `CNAME` entry is specified.

Reviewer 2: ```suggestion
. There are different `CNAME` resource records to point to depending on which {zone} you migrat the application to; please refer to xref:zones.adoc[the {zone}s] reference document, where the corresponding `CNAME` entry is specified.
```

== Let's Encrypt

> In APPUiO Cloud only `Ingress` objects can request certificates.

Reviewer: ```suggestion
In APPUiO Cloud only `Ingress` and `certificates.cert-manager.io/v1` objects can request certificates.
```

Author: Merged

IMPORTANT: It's not possible to request certificates in `Route` objects.

The manifest below specifies an `Ingress` object with such requirement:

[source,yaml]
--
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt-production

>     kubernetes.io/ingress.class: nginx

Reviewer: This annotation doesn't seem to be actually required according to https://docs.appuio.cloud/user/how-to/getting-a-certificate.html

    kubernetes.io/tls-acme: "true" # <1>
  name: ${DEPLOYMENT_NAME}

>   namespace: oauth-test

Reviewer: The `# ` isn't necessary to annotate code

> metadata:
>   annotations:
>     cert-manager.io/cluster-issuer: letsencrypt-production
>     kubernetes.io/ingress.class: nginx
>     kubernetes.io/tls-acme: "true" # <1>
>   name: ${DEPLOYMENT_NAME}
>   namespace: oauth-test
> spec:
>   rules:
>   - host: ${ROUTE_HOST}
>     http:
>       paths:
>       - backend:
>           service:
>             name: nginx
>             port:
>               number: 80
>         path: /
>         pathType: ImplementationSpecific
>   tls:
>   - hosts:
>     - ${ROUTE_HOST}
>     secretName: ${ROUTE_HOST}
> --
> <1> Annotation providing automatic provisioning of certificates on APPUiO Cloud.

Reviewer: These annotations are wrong.
First, the ingress class on Cloud is not nginx, it would be HAproxy. But I believe it's not necessary if Haproxy is the default ingress class.
Secondly, we haven't activated the `kubernetes.io/tls-acme` annotation, only the `cert-manager.io/cluster-issuer`. 
```suggestion
metadata:
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt-production <1>
  name: ${DEPLOYMENT_NAME}
  namespace: oauth-test
spec:
  rules:
  - host: ${ROUTE_HOST}
    http:
      paths:
      - backend:
          service:
            name: nginx
            port:
              number: 80
        path: /
        pathType: ImplementationSpecific
  tls:
  - hosts:
    - ${ROUTE_HOST}
    secretName: ${ROUTE_HOST}
--
<1> Annotation providing automatic provisioning of certificates on APPUiO Cloud.
```

You can find more information in the xref:how-to/getting-a-certificate.adoc[Getting a Certificate through Let’s Encrypt] page.

> For reference, it was possible to configure TLS certificates in `Route` objects in APPUiO Public, using a manifest similar to the following. This is no longer valid in APPUiO Cloud.

Reviewer: ```suggestion
For reference, it was possible to configure TLS certificates in `Route` objects in APPUiO Public, using a manifest similar to the following. 
This is no longer valid in APPUiO Cloud.
```

Author: I do not understand what changed here.

Reviewer: As  @corvus-ch already explained: We follow the one-line-per-sentence style guide in asciidoc. So I just moved `This is no longer valid in APPUiO Cloud` to the next line.

Author: We are not doing such a thing anywhere else; where is such a rule documented?

Reviewer 2: ttps://syn.tools/syn/references/style-guide.html#_asciidoc_style. Yes, I understand that that the scope of that guide and this repository do not nessesarily overlap. Hower, we from Tarazed are used to follow it and it became the defacto standard also for all we do.

[source,yaml]
--
- apiVersion: route.openshift.io/v1
  kind: Route
  metadata:
    name: ${DEPLOYMENT_NAME}
    labels:
      app: ${CI_ENVIRONMENT_SLUG}
    annotations:
      kubernetes.io/tls-acme: "${ROUTE_TLS_ACME}" # <1>
  spec:
    host: ${ROUTE_HOST}
    port:
      targetPort: http
    tls:
      insecureEdgeTerminationPolicy: Redirect
      termination: edge
    to:
        kind: Service
        name: ${DEPLOYMENT_NAME}
        weight: 100
    wildcardPolicy: None
--
<1> Annotation providing automatic provisioning of certificates on APPUiO Public.

== Storage

> APPUiO Cloud has different requirements regarding storage. This section provides more information.

Reviewer: ```suggestion
APPUiO Cloud has different requirements regarding storage.
This section provides more information.
```

=== Access the RWO volumes from different pods

When migrating your volumes to APPUiO Cloud, please follow these guidelines:

> . The rollout strategy must change from `RollingUpdate` to `Recreate`, because mounting the volume at different nodes simultaneously isn't possible, or at least not with common used filesystems, such as https://en.wikipedia.org/wiki/Ext4[ext4] and https://en.wikipedia.org/wiki/XFS[XFS]. https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#strategy[This page] in the Kubernetes documentation provides more information about this.

Reviewer: ```suggestion
. The rollout strategy must change from `RollingUpdate` to `Recreate`, because mounting the volume at different nodes simultaneously isn't possible, or at least not with commonly used filesystems, such as https://en.wikipedia.org/wiki/Ext4[ext4] and https://en.wikipedia.org/wiki/XFS[XFS]. 
  https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#strategy[This page] in the Kubernetes documentation provides more information about this.
```

> . Add an affinity rule to run on the same node as the worker pod. This allows to access the same RWO/block volume, because the filesystem is mounted on the node level, and allows a safe access to the same filesystem, because it's the same kernel.

Reviewer: ```suggestion
. Add an affinity rule to run on the same node as the worker pod. 
  This allows to access the same RWO/block volume, because the filesystem is mounted on the node level, and allows a safe access to the same filesystem, because it's the same kernel.
```

+
IMPORTANT: This could be used for jobs, but in the case of independent application parts this should be prevented.
+

> To create a affinity rule for all the pods at once, you can use the https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/[app.kubernetes.io/component label].

Reviewer 2: This is read as "an affinity". I am always uncertain if the `n` should be there in written text or not.

Reviewer: The general recommendation as I learned in university is to write `an` if the first letter is a vocal.
e.g. `an apple`, `an orange` but `a banana`

  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: edit
subjects:
- kind: ServiceAccount
  name: rsync-destination
  namespace: rsync-test
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: rsync-destination
  namespace: rsync-test
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  volumeMode: Filesystem
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/name: rsync-destination
  name: rsync-destination
  namespace: rsync-test
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: rsync-destination
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app.kubernetes.io/name: rsync-destination
    spec:
      containers:
      - image: registry.access.redhat.com/rhel7/rhel-tools
        imagePullPolicy: IfNotPresent
        name: rhel-tools
        command:
          - tail
          - -f
          - /dev/null
        volumeMounts:
        - mountPath: /rsync-destination
          name: rsync-destination
      volumes:
      - name: rsync-destination
        persistentVolumeClaim:
          claimName: rsync-destination
--

=== Job-Based Migration

> The second option for migrating your storage from APPUiO Public to APPUiO Cloud consists in using jobs. The manifest below defines the objects required for this strategy.

Reviewer: ```suggestion
The second option for migrating your storage from APPUiO Public to APPUiO Cloud consists in using jobs.
The manifest below defines the objects required for this strategy.
```

Author: Again, what's the purpose of these changes?

              name: rsync-source
          restartPolicy: Never
          volumes:
          - name: rsync-source
            persistentVolumeClaim:
              claimName: rsync-source
  schedule: '@yearly'
  startingDeadlineSeconds: 86400
  successfulJobsHistoryLimit: 1
--
<1> Please refer to https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/[the Kubernetes documentation] on common labels.
<2> Use `batch/v1` for OpenShift 4 instead.

Use the commands below to create a new job based on the definition above:

[source,bash]
--
$ JOB_NAME="manual-$(date +%F-%H-%M)" oc -n rsync-test create job --from=cronjob/rsync-copy $JOB_NAME

$ oc -n rsync-test get po
NAME                                 READY   STATUS      RESTARTS   AGE
manual1-8975l                        0/1     Completed   0          2m9s
rsync-destination-6fd76657d8-6fjss   1/1     Running     0          41m
rsync-source-957bf555c-68jmn         1/1     Running     0          5m5s

$ oc -n rsync-test delete job $JOB_NAME
--

Check the job status with the following command:

[source,bash]
--
$ oc -n <namespace> get job <myjob> -o jsonpath={.status.succeeded}
--

### Continuous Sync

> The third option to migrate your storage to APPUiO Cloud consists in using a Continuous Sync strategy. The main benefit of this approach is that files are replicated immediately after they're created. If the destination pod dies, the sync pod also crashes, but is automatically restarted.

Reviewer: ```suggestion
The third option to migrate your storage to APPUiO Cloud consists in using a Continuous Sync strategy. The main benefit of this approach is that files are replicated immediately after they're created.
If the destination pod dies, the sync pod also crashes, but is automatically restarted.
```

 -c, --container='': Container within the pod
     --delete=false: If true, delete files not present in source
     --exclude=[]: When specified, exclude files matching pattern
     --include=[]: When specified, include files matching pattern
     --no-perms=false: If true, do not transfer permissions
     --progress=false: If true, show progress during transfer
 -q, --quiet=false: Suppress non-error messages
     --strategy='': Specify which strategy to use for copy: rsync, rsync-daemon, or tar
 -w, --watch=false: Watch directory for changes and resync automatically
--

=== File Integrity Check

After the migration, you should check the integrity of your data with the following commands:

.Calculate the checksum for all files at the origin
[source,bash]
--
$ find . -type d -exec sh -c "cd '{}' && find . -maxdepth 1 -type f ! -name COPYSHA1SUMS -printf '%P\0' | xargs -r0 sha1sum -- > COPYSHA1SUMS" \;
--

.Create a log with a verification of all files at the destination
[source,bash]
--
$ cd <path> && find . -type d -exec sh -c "cd '{}' && echo '{}' && sha1sum -c COPYSHA1SUMS" \; > sha1sums-verify-log-$(date +%F-%H-%M).log 2>&1
--

== Container Images

> Since APPUiO Cloud is based on OpenShift 4, there are new requirements for your container images. This section contains all the required steps for adapting your images to the new environment.

Reviewer: ```suggestion
Since APPUiO Cloud is based on OpenShift 4, there are new requirements for your container images.
This section contains all the required steps for adapting your images to the new environment.
```

=== On OpenShift 3

TIP: This section uses the https://github.com/containers/skopeo[skopeo] tool for managing images and repositories.

. Get a user token at this URL: https://<origin-cluster-console>/oauth/token/request.

> . Use the generated user token to authenticate to the registry on the command line. As the user token has enough privileges to read the image, a service account token isn't required.

Reviewer: ```suggestion
. Use the generated user token to authenticate to the registry on the command line.
  As the user token has enough privileges to read the image, a service account token isn't required.
```

+
[source,bash]
--
$ skopeo login -u openshift -p <token> <origin-url>
--
+
Skopeo uses the docker auth config. So this should look like:
+
[source,bash]
--
$ cat ~/.docker/config.json
{
    "auths": {
        "<origin-url>": {
            "auth": "...="
        }
    }
}
--
+
Check if the access is working:
+
[source,bash]
--
$ skopeo inspect docker://<origin-url>/<namespace>/<image>:<image-tag>
--

IMPORTANT: This is a user token, and therefore it expires when you logout.

=== On OpenShift 4

> On OpenShift 4 it's also possible to find the token from https://oauth-openshift.apps.<cluster>/oauth/token/display and get read access; but this token doesn't grant enough privileges to write images. Therefore it's recommended to create a service account, and to grant access to `system:image-builders`, and finally to get the token from this service account.

Reviewer: ```suggestion
On OpenShift 4 it's also possible to find the token from https://oauth-openshift.apps.<cluster>/oauth/token/display and get read access; but this token doesn't grant enough privileges to write images.
  Therefore it's recommended to create a service account, and to grant access to `system:image-builders`, and finally to get the token from this service account.
```

        weight: 100
    wildcardPolicy: None
--
<1> Annotation providing automatic provisioning of certificates on APPUiO Public.

== Storage

APPUiO Cloud has different requirements regarding storage.
This section provides more information.

=== Access the RWO volumes from different pods

When migrating your volumes to APPUiO Cloud, please follow these guidelines:

. The rollout strategy must change from `RollingUpdate` to `Recreate`, because mounting the volume at different nodes simultaneously isn't possible, or at least not with commonly used filesystems, such as https://en.wikipedia.org/wiki/Ext4[ext4] and https://en.wikipedia.org/wiki/XFS[XFS]. 
  https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#strategy[This page] in the Kubernetes documentation provides more information about this.

. Add an affinity rule to run on the same node as the worker pod. This allows to access the same RWO/block volume, because the filesystem is mounted on the node level, and allows a safe access to the same filesystem, because it's the same kernel.
+
IMPORTANT: This could be used for jobs, but in the case of independent application parts this should be prevented.
+

> To create a affinity rule for all the pods at once, you can use the https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/[app.kubernetes.io/component label].

Reviewer 3: ```suggestion
To create an affinity rule for all the pods at once, you can use the https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/[app.kubernetes.io/component label].
```

. Add an affinity rule to run on the same node as the worker pod. 
  This allows to access the same RWO/block volume, because the filesystem is mounted on the node level, and allows a safe access to the same filesystem, because it's the same kernel.
+
IMPORTANT: This could be used for jobs, but in the case of independent application parts this should be prevented.
+
To create a affinity rule for all the pods at once, you can use the https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/[app.kubernetes.io/component label].
+
[source,yaml]
--
affinity:
  podAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
    - labelSelector:
        matchExpressions:
        - key: app.kubernetes.io/component
           operator: In
          values:
          - backend
      topologyKey: kubernetes.io/hostname
--

. Perform the migration itself. There are three major mechanisms for this:
.. Migration with `rsync`
.. Migration through jobs
.. Using continuous sync

The following sections provide information about each strategy.

### Migration with `rsync`

> The first option to migrate your storage from APPUiO Public to APPUiO Cloud consists in using `rsync`. The manifests below create the required objects.

Reviewer 3: ```suggestion
The first option to migrate your storage from APPUiO Public to APPUiO Cloud consists in using `rsync`.
The manifests below create the required objects.
This part has to be installed on the destination side.
Overall a pod must be running with the RWO volume mounted, the `oc rsync` can connect to.
```

= Migration Tutorial: From APPUiO Public to APPUiO Cloud

This document provides an overview of all the tasks required to move your applications from APPUiO Cloud to APPUiO Public.

== Main Differences

Before starting the migration of your applications from APPUiO Public to APPUiO Cloud, please check the xref:differences-to-public.adoc[document explaining all the major differences] between both platforms.

> === Timezone

Reviewer 4: I don't think that the Timezone is the most important difference to especially highlight here. I'd rather improve the `differences-to-public.adoc` regarding that topic.

Author: Removed

As a reminder, please keep in mind that in APPUiO Public the timezone was *Europe/Zurich*; however, in APPUiO Cloud (based on OpenShift 4) the control plane runs on *UTC* instead.

Example of getting the current time on a control plane:

[source,bash]
--
sh-4.4# date
Mon Oct 25 14:41:20 UTC 2021
--

> For all time and date information in your manifests and commands, please change it to conform to UTC, or better yet, beginning with OpenShift 4.9 (based on Kubernetes v1.21) you should define the timezone in a https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/[Kubernetes CronJob].

Reviewer 4: 4.9 would be 1.22 (https://docs.openshift.com/container-platform/4.9/release_notes/ocp-4-9-release-notes.html#ocp-4-9-about-this-release)

Author: Section removed

== DNS

Customers transferring their domains from APPUiO Public to APPUiO Cloud should be aware of the following:

> . Lowering your TTLs (Time To Live) 24 hours ahead of time will reduce the amount of time that your change takes to propagate. You can limit downtime by lowering that value, so the caching servers will check back more frequently. Typically, 300 seconds (5 minutes) instead of the default of 14400 (4 hours) is a good idea.

Reviewer 4: ```suggestion
. Lowering your TTLs (Time To Live) 24 hours ahead of time will reduce the amount of time that your change takes to propagate. You can limit downtime by lowering that value, so the caching servers will check back more frequently. Typically, 60 seconds (1 minute) is a good idea.
```

Note: There is no default of TTL.

Author: Merged

+

> [source,bash]

Reviewer 4: I'm not sure how much value this example brings to this stage, I'd just remove it.

Author: Removed

--
$ vi /var/named/example.com.db # <1>
$ rndc reload example.com # <2>
$ dig @localhost example.com # <3>
--
<1> Edit the zone file
<2> Make BIND aware of your changes
<3> Test your changes using https://www.liquidweb.com/kb/how-to-use-dig/[dig]
+
IMPORTANT: Remember to set your TTL back to the previous value after the migration.

> . There are different `CNAME` to point to depending on your current zone; please refer to xref:zones.adoc[the APPUiO Cloud Zones] reference document, where the corresponding `CNAME` entry is specified.

Reviewer 4: ```suggestion
. There are different `CNAME` resource records to point to depending on which {zone} you migrated the application to; please refer to xref:zones.adoc[the {zone}s] reference document, where the corresponding `CNAME` entry is specified.
```

Author: Changed

            name: nginx
            port:
              number: 80
        path: /
        pathType: ImplementationSpecific
  tls:
  - hosts:
    - ${ROUTE_HOST}
    secretName: ${ROUTE_HOST}
--
<1> Annotation providing automatic provisioning of certificates on APPUiO Cloud.

For reference, it was possible to configure TLS certificates in `Route` objects in APPUiO Public, using a manifest similar to the following. This is no longer valid in APPUiO Cloud.

[source,yaml]
--
- apiVersion: route.openshift.io/v1
  kind: Route
  metadata:
    name: ${DEPLOYMENT_NAME}
    labels:
      app: ${CI_ENVIRONMENT_SLUG}
    annotations:
      kubernetes.io/tls-acme: "${ROUTE_TLS_ACME}" # <1>
  spec:
    host: ${ROUTE_HOST}
    port:
      targetPort: http
    tls:
      insecureEdgeTerminationPolicy: Redirect
      termination: edge
    to:
        kind: Service
        name: ${DEPLOYMENT_NAME}
        weight: 100
    wildcardPolicy: None
--
<1> Annotation providing automatic provisioning of certificates on APPUiO Public.

== Storage

APPUiO Cloud has different requirements regarding storage. This section provides more information.

=== Access the RWO volumes from different pods

> On APPUiO Public the default was RWX storage. At the moment, *APPUiO Cloud only features RWO storage*. Former deployments using RWX volumes can't run anymore without modification.

Reviewer 4: As we'll be enabling RWX storage in the next two weeks, I'd rather not mention that here as it will be outdated very soon (and the "Ready for Migration" milestone - which this documentation targets - will be in a month).

Author: Removed the reference

. Add an affinity rule to run on the same node as the worker pod. This allows to access the same RWO/block volume, because the filesystem is mounted on the node level, and allows a safe access to the same filesystem, because it's the same kernel.
+
IMPORTANT: This could be used for jobs, but in the case of independent application parts this should be prevented.
+
To create a affinity rule for all the pods at once, you can use the https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/[app.kubernetes.io/component label].
+
[source,yaml]
--
affinity:
  podAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
    - labelSelector:
        matchExpressions:
        - key: app.kubernetes.io/component
           operator: In
          values:
          - backend
      topologyKey: kubernetes.io/hostname
--

. Perform the migration itself. There are three major mechanisms for this:
.. Migration with `rsync`
.. Migration through jobs
.. Using continuous sync

The following sections provide information about each strategy.

### Migration with `rsync`

> The first option to migrate your storage from APPUiO Public to APPUiO Cloud consists in using `rsync`. The manifests below create the required objects.

Reviewer 4: This might need some more explanation: Where do I create these objects?

Author: I'll wait for the review by @megian to rewrite this, as this section comes directly from the text of his ticket #36

Reviewer 3: This part has to be installed on the destination side. Overall a pod must be running with the RWO volume mounted, the òc rsync` can connect to.

--

== Container Images

Since APPUiO Cloud is based on OpenShift 4, there are new requirements for your container images. This section contains all the required steps for adapting your images to the new environment.

=== On OpenShift 3

TIP: This section uses the https://github.com/containers/skopeo[skopeo] tool for managing images and repositories.

. Get a user token at this URL: https://<origin-cluster-console>/oauth/token/request.

. Use the generated user token to authenticate to the registry on the command line. As the user token has enough privileges to read the image, a service account token isn't required.
+
[source,bash]
--
$ skopeo login -u openshift -p <token> <origin-url>
--
+
Skopeo uses the docker auth config. So this should look like:
+
[source,bash]
--
$ cat ~/.docker/config.json
{
    "auths": {
        "<origin-url>": {
            "auth": "...="
        }
    }
}
--
+
Check if the access is working:
+
[source,bash]
--
$ skopeo inspect docker://<origin-url>/<namespace>/<image>:<image-tag>
--

IMPORTANT: This is a user token, and therefore it expires when you logout.

> === On OpenShift 4

Reviewer 4: This part might need APUB-72 which is scheduled for the next sprint (aka next week).

Author: Should I add a reference to this, or just leave it like that, since it should be something available by the time we publish this?

Reviewer 4: No ref needed, just leave it as it is.

Author: Perfect! Resolving this conversation.
